### ISSUE TITLE
Interpolation bug for macro-averaged multiclass precision-recall curves (`torchmetrics.functional.classification.precision_recall_curve._multiclass_precision_recall_curve_compute`)

### ISSUE THREAD

**Issue by EdMorrell** [Lightning-AI/torchmetrics#3413]
## 🐛 Bug

When computing a macro-averaged precision-recall curve, [`_multiclass_precision_recall_curve`](https://github.com/Lightning-AI/torchmetrics/blob/master/src/torchmetrics/functional/classification/precision_recall_curve.py#L537) interpolates recall using precision values as `xp`, however [`interp`](https://github.com/Lightning-AI/torchmetrics/blob/master/src/torchmetrics/utilities/compute.py#L168) requires `xp` to be monotonically increasing, and precision is not guaranteed to be monotonic which can lead to weird-looking plots, especially when predictions are poorly correlated with the target.

**Micro-Average**
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/d7619b74-113a-4516-b4dc-87d39ffcf6e7" />

**Macro-Average**
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/c7dee419-5381-4d95-9d5f-fb719a5183d3" />

#### Contrived Example
See the code, below. Here I deliberately generated data where precision doesn't vary monotonically to demonstrate how this leads to weird non-monotonic PR curves for `macro` averaging.

**Micro-Average**

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/495ad0ac-3f2e-4c83-8dcf-4f6848ce1ba3" />

**Macro-Average**

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/cc51a424-af72-4b04-b7bc-61073698b79e" />

### Proposed Fix
The fix is very straightforward. We just need to swap round the interpolation (so we interpolate precision, using recall values rather than vice versa.) I'd be more than happy to submit a (draft) PR with an implementation if this would be helpful?

**Macro-Average After Fix (for first plot)**

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/eeb56174-3ae6-4e16-a54c-633453dfa161" />

(_Note:_ Recall is now monotonic as is typical for a precision-recall curve)

**Macro-Average After Fix (for contrived example)**

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/a51b86b5-ceef-4668-a6a8-c34b1afea050" />

### To Reproduce

Sample code and steps to reproduce the behavior with expected result...

<details>
  <summary>Code sample for Semi-Correlated Data</summary>

```python
import torch
import matplotlib.pyplot as plt
import torch.nn.functional as F
from torchmetrics.classification import MulticlassPrecisionRecallCurve

torch.manual_seed(42)

n_samples = 250
n_classes = 3
average = "micro"  # Swap to view the 2 different curve-types

target = torch.randint(
    low=0,
    high=n_classes,
    size=(n_samples,)
)

logits = torch.randn(n_samples, n_classes)

# Boost the true class so predictions are correlated with labels
boost = 0.1
logits[torch.arange(n_samples), target] += boost

preds = F.softmax(logits, dim=1)


pr_curve = MulticlassPrecisionRecallCurve(num_classes=n_classes, average=average)
pr_curve.update(preds, target)
curve = pr_curve.plot()
plt.show()
```
</details>

<details>
    <summary>Code-Sample for Contrived Example</summary>

```python
import torch
import matplotlib.pyplot as plt
import torch.nn.functional as F
from torchmetrics.classification import MulticlassPrecisionRecallCurve

torch.manual_seed(42)

n_samples = 1000
n_classes = 3
average = "micro"

preds = torch.zeros(12, n_classes)
target = torch.zeros(12, dtype=torch.long)

# Block 1 - High confidence but wrong -> High FPs
for i in range(3):
    preds[i] = torch.tensor([0.85 - i * 0.01, 0.10, 0.05])
    target[i] = 1

# Block 2 - Medium confidence, correct -> Precision recovers
for i in range(3, 6):
    preds[i] = torch.tensor([0.6 - i * 0.01, 0.25, 0.15])
    target[i] = 0

# Block 3 - Lower confidence, incorrect -> Precision drops
for i in range(6, 9):
    preds[i] = torch.tensor([[0.45 - (i - 6) * 0.01, 0.20, 0.35]])
    target[i] = 2

# Block 4 - Lowest confidence, correct -> Precision recovers
for i in range(9, 12):
    preds[i] = torch.tensor([[0.35 - (i - 9) * 0.01, 0.30, 0.35]])
    target[i] = 2


pr_curve = MulticlassPrecisionRecallCurve(num_classes=n_classes, average=average)
pr_curve.update(preds, target)
curve = pr_curve.plot()
plt.show()

```

</detail>

<details>
  <summary>Environment</summary>

- TorchMetrics version (if build from source, add commit SHA): 1.9.0
- Python & PyTorch Version (e.g., 1.0): 3.14.5
- Any other relevant information such as OS (e.g., Linux): MacOS

</details>

### Additional context

N/A

**Comment by EdMorrell** [comment]
I have opened a draft PR to fix this issue [#3414](https://github.com/Lightning-AI/torchmetrics/pull/3414). If you're happy with the implementation I can convert to a full PR?

**Comment by hiSandog** [comment]
The contrived example is a good basis for a regression test because it isolates the non-monotonic precision axis. The fix should probably assert two things: interpolation uses recall or another monotonic support as `xp`, and the macro curve remains ordered/plot-safe even when per-class precision is noisy. That would protect both the numeric result and the user-visible weird plot shown in the issue.

**Comment by EdMorrell** [comment]
I've added a dedicated test for this specific case (non-monotonic precision) in the draft PR shared above ([#3414](https://github.com/Lightning-AI/torchmetrics/pull/3414)). I've hardcoded the specific test-case input as it's a bit tricky to generate parametrised inputs that would have this property without quite complex test logic but let me know if you think another approach would work better.

In terms of the assertions, this test verifies non-monotonicity of the precision axis (specifically designed by the test to be like this), and monotonicity of the recall axis, which also implicitly verifies that the plot will visually look as expected. In terms of verifying that interpolation uses the correct axis, I'm not sure how to do this without patching that function, but that doesn't seem to be a common pattern in this codebase, so left if at just these assertions (verifying expected output for a given input). However let me know if you think an alternative approach would work better here? One approach might be to hard-code the expected output as well?

I've also tested this with both the old implementation and the new implementation and verified it fails with the former and passes with the latter.
